File: Code/Sandbox/Old/TF_VMC_notf_old/WaveFunctions/simplegaussian.py
import numpy as np
from particles import Particles   
from WaveFunctions.wavefunction import WaveFunction

class SimpleGaussian(WaveFunction):
    def __init__(self, alpha, HOshape, particles):
        self._alpha = alpha
        self._HOs = HOshape
        self._num = particles
    
    def kinetic(self, particles):
        return self._alpha * np.sum(self._HOs - 2 * self._alpha * np.power(particles.getPositions() * self._HOs, 2))
    
    def QF(self, particles, p1):
        return -4 * self._alpha * particles.getPos(p1) * self._HOs
    
    def ratio(self, particles, p1, oldPos):
        """
        Old wavefunc value term: exp(-alpha * old_r2)
        New wavefunc value term: exp(-alpha * r2)
        All other terms are the same and cancel. Since they are both exponentials, we can subtract the exponents
        """
        r2 = np.sum(particles.getPos(p1) * particles.getPos(p1) * self._HOs)
        old_r2 = np.sum(oldPos * oldPos * self._HOs)
        return np.exp(-self._alpha * r2 + self._alpha * old_r2)
    
    def paramDer(self, positions):
        return -np.sum(positions * positions * self._HOs) / self._num

File: Code/Sandbox/Old/RBM/RBM_VMC/Wavefunctions/simplegaussian.jl
struct SimpleGaussian <: WaveFunction
    alpha::Float64
    HOshape::Vector{Float64}
    HOshape2::Vector{Float64}
    function SimpleGaussian(alpha, HOshape)
        return new(alpha, HOshape, HOshape.^2)
    end
end


function evaluate(particles, wf::SimpleGaussian)::Float64
    """
    Old wavefunc value term: exp(-alpha * old_r2)
    New wavefunc value term: exp(-alpha * r2)
    All other terms are the same and cancel. Since they are both exponentials, we can subtract the exponents
    """
    exponent = 0.0
    for d in 1:particles.dims
        for i in d:particles.dims:length(particles.positions)
            exponent += particles.positions[i]^2 * wf.HOshape[d]
        end
    end
    return exp(-wf.alpha * exponent)
end

function kinetic(particles, wf::SimpleGaussian)::Float64
    kin_sum = 0.0
    for d in 1:particles.dims
        for i in d:particles.dims:length(particles.positions)
            kin_sum += particles.positions[i]^2 * wf.HOshape2[d]
        end
    end
    return wf.alpha * (sum(wf.HOshape) * particles.num - 2.0 * wf.alpha * kin_sum)
end

function QF(particles, idx, wf::SimpleGaussian) #maybe faster to put return type here?
    @views qf = -4.0 * wf.alpha .* particles.positions[idx] .* wf.HOshape
    return qf
end

function QF!(qf, particles, idx, wf::SimpleGaussian) #maybe faster to put return type here?
    @views qf .= -4.0 * wf.alpha .* particles.positions[idx] .* wf.HOshape
    return qf
end

function paramDer(particles, wf::SimpleGaussian)::Float64
    pos_sum = 0.0
    for d in 1:particles.dims
        for i in d:particles.dims:length(particles.positions)
            pos_sum += particles.positions[i]^2 * wf.HOshape[d]
        end
    end
    return -pos_sum / particles.num
end

function applyGradient(wf::SimpleGaussian, grad)
    return SimpleGaussian(wf.alpha - grad, wf.HOshape)
end